File: Problems/ICPC/ReportRecovery.py

from pprint import pprint

N = None
P = None
names = []
numbers = []
totals = None
pos = None

def check1(p, curr):
    s = 0
    l = 0
    for i in curr[:-1]:
        s += int(numbers[p][l : i])
        l = i
    if s == int(numbers[p][l : curr[-1]]):
        return True
    return False

def generate_pos(p, i, n, curr):
    if n > N + 1:
        return
    if i == len(numbers[p]):
        if n == N + 1:
            if check1(p, curr.copy()):
                pos[p].append(curr.copy())
        return

    for j in range(i, max(i + 1, int(numbers[p][i] != '0') * len(numbers[p]))):
        curr.append(j + 1)
        generate_pos(p, j + 1, n + 1, curr.copy())
        curr = curr[:-1]

def check2(curr):
    l = [0 for _ in range(P + 1)]
    for i in range(N + 1):
        s = 0
        for j in range(P):
            s += int(numbers[j][l[j] : pos[j][curr[j]][i]])
            l[j] = pos[j][curr[j]][i]
        if s != int(numbers[-1][l[-1] : pos[-1][curr[-1]][i]]):
            return False
        l[-1] = pos[-1][curr[-1]][i]
    return True

def final(p, curr):
    if p == P + 1:
        if check2(curr.copy()):

            print("gg")

            return True
        return False

    for i in range(len(pos[p])):
        curr.append(i)
        if final(p + 1, curr.copy()):
            return True
        curr = curr[:-1]
    return False

if __name__ == "__main__":

    T = int(input())

    for t in range(T):

        N = len(input().strip().split('P')[1:])
        numbers = []
        names = []

        while True:

            line = input().strip()
            name = ""
            nums = ""

            for l in line:
                if l.isalpha():
                    name += l
                else:
                    nums += l

            numbers.append(nums)
            names.append(name)

            if name == "TP":
                break

        P = len(names) - 1

        pos = [[] for i in range(P + 1)]

        for i in range(P + 1):
            generate_pos(i, 0, 0, [])

        final(0, [])

        pprint(pos)

File: Problems/ICPC/ReportRecovery.cpp
#include <bits/stdc++.h>
using namespace std;

#define rep(i, n) for (int i = 0; i < (int)n; i++)
#define repx(i, a, b) for (int i = (int)a; i < (int)b; i++)

int T, N, P;
string line, name, num;
vector<string> names, numbers;
vector<vector<vector<int>>> pos;
vector<int> curr;

bool check1(int p)
{
    int s = 0;
    int l = 0;
    rep(i, N)
    {
        if (curr[i] - l > 6) return false;
        s += stoi(numbers[p].substr(l, curr[i] - l));
        l = curr[i];
    }
    if (curr[N] - l > 6) return false;
    if (s == stoi(numbers[p].substr(l, curr[N] - l)))
        return true;
    return false;
}

void generate_pos(int p, int i, int n)
{
    if (n > N + 1) return;
    if (i == numbers[p].size())
    {
        if (n == N + 1 and check1(p)) pos[p].push_back(curr);
        return;
    }

    repx(j, i, max(i + 1, int(numbers[p][i] != '0') * (int)numbers[p].size()))
    {
        curr.push_back(j + 1);
        generate_pos(p, j + 1, n + 1);
        curr.pop_back();
    }
}

bool check2()
{
    vector<int> l(P + 1, 0);
    rep(i, N + 1)
    {
        int s = 0;
        rep(j, P)
        {
            s += stoi(numbers[j].substr(l[j], pos[j][curr[j]][i] - l[j]));
            l[j] = pos[j][curr[j]][i];
        }
        if (s != stoi(numbers[P].substr(l[P], pos[P][curr[P]][i] - l[P])))
            return false;
        l[P] = pos[P][curr[P]][i];
    }
    return true;
}

bool final(int p)
{
    if (p == P + 1)
    {
        if (check2())
        {
            rep(i, N) cout << 'P' << i + 1 << ' ';
            cout << "Totals\n";

            rep(i, P + 1)
            {
                cout << names[i] << ' ';
                int l = 0;
                rep(j, N)
                {
                    cout << numbers[i].substr(l, pos[i][curr[i]][j] - l) << ' ';
                    l = pos[i][curr[i]][j];
                }
                cout << numbers[i].substr(l, pos[i][curr[i]][N] - l) << '\n';
            }
            return true;
        }
        return false;
    }

    rep(i, pos[p].size())
    {
        curr.push_back(i);
        if (final(p + 1)) return true;
        curr.pop_back();
    }
    return false;
}

int main()
{
    cin >> T;
    while (T--)
    {
        names.clear();
        numbers.clear();

        cin >> line;
        N = ((int)line.size() - 6) / 2;

        while (true)
        {
            cin >> line;

            name = num = "";
            for (char c : line)
            {
                if (c < 'A') num += c;
                else name += c;
            }

            names.push_back(name);
            numbers.push_back(num);

            if (name == "TP") break;
        }

        P = names.size() - 1;

        pos.assign(P + 1, {});
        rep(i, P + 1) generate_pos(i, 0, 0);

        final(0);
        curr.clear();
    }
}